src/rabbit_channel_interceptor_new.erl

```diff
@@ -0,0 +1,67 @@
+-module(rabbit_channel_interceptor_new).
+
+-include("rabbit_framing.hrl").
+-include("rabbit.hrl").
+
+-export([init/1, intercept_in/3]).
+
+-define(OLD_REGISTRY_CLASS, channel_interceptor).
+-define(NEW_REGISTRY_CLASS, channel_interceptor_new).
+
+-ifdef(use_specs).
+
+% TODO: types
+-callback init(any) -> any.
+-callback intercept(any, any, any) -> any.
+-callback applies_to(any) -> any.
+
+-else.
+
+% TODO
+
+-endif.
+
+init(Ch) ->
+  R = (catch begin
+    Mods = lookup_all(?OLD_REGISTRY_CLASS) ++ lookup_all(?NEW_REGISTRY_CLASS),
+    {Hot, Cold} = lists:partition(fun ({_, M}) -> is_hot(M) end, Mods),
+    {to_fn(Hot, Ch), to_fn(Cold, Ch)}
+  end),
+  %io:format("R=~p~n", [R]),
+  R.
+
+lookup_all(Class) ->
+  [{Class, M} || {_, M} <- rabbit_registry:lookup_all(Class)].
+
+to_fn(Mods, Ch) ->
+  case Mods of
+    [] -> fun(M, C) -> {M, C} end;
+    [Mod] -> to_fn1(Mod, Ch)
+  end.
+
+to_fn1({channel_interceptor, Mod}, Ch) ->
+  VHost = rabbit_channel:get_vhost(Ch),
+  fun (M, C) -> {Mod:intercept(M, VHost), C} end;
+
+to_fn1({channel_interceptor_new, Mod}, Ch) ->
+  St = Mod:init(Ch),
```

Comment:
`Mod:init` is an interceptor callback right? What is expected of said function? What state does it tracks?